PyDSS/pyPostprocessor/PostprocessScripts/AutomatedVoltageUpgrade.py

```diff
@@ -1,1855 +1,2318 @@
+                self.max_regs = int(min(self.config["max_regulators"], len(self.buses_with_violations)))
+                self.get_shortest_path()
+                self.get_full_distance_dict()
+                self.cluster_square_array()
+                min_severity = pow(len(self.all_bus_names), 2) * len(self.config["tps_to_test"]) * self.upper_limit
+                #  determining key with minimum objective func. at various levels
+                # (at this point includes pre-reg, sub-LTC, and all newly added regulators)
+                min_cluster = ''
+                for key, vals in self.cluster_optimal_reg_nodes.items():
+                    if vals[0] < min_severity:
+                        min_severity = vals[0]
+                        min_cluster = key
+                # Logic is if violations were less before addition of any device, revert back to that condition by removing
+                #  added LTC and not adding best determined in-line regs, else if LTC was best - pass and do not add new
+                #  in-line regs, or if some better LTC placemenr was determined apply that
+                self.logger.info(f"Checking objective function to determine best possible upgrades.\n"
+                                 f"At stages: 1) Before addition of new devices. "
+                                 f"2) With Substation LTC. 3) With newly added regulators")
+                self.compare_objective_function(min_cluster)
+
+                self.check_voltage_violations_multi_tps(upper_limit=self.upper_limit,
+                                                        lower_limit=self.lower_limit)
+                self.logger.info("Compare objective with best and applied settings, %s, %s", self.cluster_optimal_reg_nodes[min_cluster][0],
+                      self.severity_indices[2])
+                self.logger.info("Additional regctrl  devices: %s", min_cluster)
+                self.logger.info("cluster_optimal_reg_nodes=%s", self.cluster_optimal_reg_nodes)
+
+        end_t = time.time()
+        # self.check_voltage_violations_multi_tps(upper_limit=self.upper_limit,
+        #                                         lower_limit=self.lower_limit)
+        # if self.config["create_topology_plots"]:
+        #     self.plot_violations()
+        self.logger.debug("Writing upgrades to DSS file")
+        self.write_upgrades_to_file()
+        self.logger.info("total_time = %s", end_t - start_t)
+
+        # TODO: Check impact of upgrades - Cannot recompile feeder in PyDSS
+        self.logger.info("Checking impact of redirected upgrades file")
+        dss.run_command("Clear")
+        base_dss = os.path.join(project.dss_files_path, self.Settings["Project"]["DSS File"])
+        # check_redirect(base_dss)
+        result = dss.run_command(f"Redirect {base_dss}")
+        if result != "":
+            print(f"Redirect failed for {base_dss}, message: {result}")

```

Comment:
Should this raise an exception?  The file did not compile.  check_direct above?

Reply:
Code was repeated. 
check_redirect does the same action. Fixed.